### DISCUSSION TITLE
How can async callbacks send events into the machine?

### DISCUSSION THREAD

**Discussion by markmelville** [statelyai/xstate#1330] [General] [answered]
I have a method that calls an API which sends back data as server-sent events (SSE). The method signature is `openEventStream(messageCallback)`. The server returns 200 immediately and the SSE messages are received via the callback. In a machine, the early Promise fulfillment transitions to `processing` where it should then add each received message into the context. When a falsy message comes in, it should transition to `finish`:
```
import { Machine, assign, send } from "xstate";

const progress = Machine({
  id: "sse",
  initial: "start",
  context: {
    messages: []
  },
  states: {
    start: {
      invoke: {
        id: 'getProgress',
        src: (context, event) => openEventStream(message => send(message?"PROGRESS":"DONE", message)),
        onDone: 'processing',
        onError: 'failed'
      }
    },
    processing: {
      on: {
        "PROGRESS": {
          actions: assign({
            messages: (context,event)=>[...context.messages,event]
          })
        },
        "DONE": "finish",
      }
    },
    finish: {},
    failed: {}
  }
});
```
By using advanced `console.log` debugging technology, I have determined that the messageCallback is invoked, but the PROGRESS assign function is never invoked, nor is the machine ever transitioned to `finish`. Why are events not making it into the machine? Is there a more correct way to do this?

UPDATE: I created a codesandbox for it. You can see the events come through in the console, but the messages never make it into the context to be rendered.
https://codesandbox.io/s/inspiring-lalande-v8ymg

**Comment by davidkpiano** [comment, marked as answer]
You're using `send(...)`, which is an **action creator** (pure function) and doesn't magically do anything. You want to use the first argument (which I called `sendBack` here) instead: https://codesandbox.io/s/stupefied-bell-vm8l8?file=/src/App.js

**Reply by markmelville** [reply]
Thank you! I updated mine based on your solution, and made a slight alteration where the promise from openEventStream is returned so its fulfillment can trigger the state transition. Plus I got the messages rendering as desired. Thanks again. I'm loving XState for UI!
